toolchain/lex/lex.cpp

```diff
@@ -0,0 +1,1309 @@
+
+#include <array>
+
+#include "common/check.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/ADT/StringSwitch.h"
+#include "llvm/Support/Compiler.h"
+#include "toolchain/base/value_store.h"
+#include "toolchain/lex/character_set.h"
+#include "toolchain/lex/helpers.h"
+#include "toolchain/lex/numeric_literal.h"
+#include "toolchain/lex/string_literal.h"
+#include "toolchain/lex/tokenized_buffer.h"
+
+#if __ARM_NEON
+#include <arm_neon.h>
+#define CARBON_USE_SIMD 1
+#elif __x86_64__
+#include <x86intrin.h>
+#define CARBON_USE_SIMD 1
+#else
+#define CARBON_USE_SIMD 0
+#endif
+
+namespace Carbon::Lex {
+
+// Implementation of the lexer logic itself.
+//
+// The design is that lexing can loop over the source buffer, consuming it into
+// tokens by calling into this API. This class handles the state and breaks down
+// the different lexing steps that may be used. It directly updates the provided
+// tokenized buffer with the lexed tokens.
+//
+// Because of the internal linkage attribute, this must be in the cpp file where
+// it's used.
+// TODO: Why is this needed?
+class [[clang::internal_linkage]] Lexer {
+ public:
+  // Symbolic result of a lexing action. This indicates whether we successfully
+  // lexed a token, or whether other lexing actions should be attempted.
+  //
+  // While it wraps a simple boolean state, its API both helps make the failures
+  // more self documenting, and by consuming the actual token constructively
+  // when one is produced, it helps ensure the correct result is returned.
+  class LexResult {
+   public:
+    // Consumes (and discard) a valid token to construct a result
+    // indicating a token has been produced. Relies on implicit conversions.
+    // NOLINTNEXTLINE(google-explicit-constructor)
+    LexResult(Token /*discarded_token*/) : LexResult(true) {}
+
+    // Returns a result indicating no token was produced.
+    static auto NoMatch() -> LexResult { return LexResult(false); }
+
+    // Tests whether a token was produced by the lexing routine, and
+    // the lexer can continue forming tokens.
+    explicit operator bool() const { return formed_token_; }
+
+   private:
+    explicit LexResult(bool formed_token) : formed_token_(formed_token) {}
+
+    bool formed_token_;
+  };
+
+  Lexer(SharedValueStores& value_stores, SourceBuffer& source,
+        DiagnosticConsumer& consumer)
+      : buffer_(value_stores, source),
+        consumer_(consumer),
+        translator_(&buffer_),
+        emitter_(translator_, consumer_),
+        token_translator_(&buffer_),
+        token_emitter_(token_translator_, consumer_) {}
+
+  // Find all line endings and create the line data structures.
+  //
+  // Explicitly kept out-of-line because this is a significant loop that is
+  // useful to have in the profile and it doesn't simplify by inlining at all.
+  // But because it can, the compiler will flatten this otherwise.
+  LLVM_ATTRIBUTE_NOINLINE auto CreateLines(llvm::StringRef source_text) -> void;
```

Comment:
I'm OK using the LLVM macros here, but we don't in other places where our pinned compiler environment removes the portability need. I have a mild preference for just directly using the C++ attribute syntax when available and we don't need any preprocessor-level portability steps.

Still, this is really trivial to switch whichever direction and as many times as we like so happy for it to be a fix-forward.

Reply:
I've gone back to gnu::noinline. I'm a little confused though: we definitely use a mix with things like LLVM_DUMP_METHOD which is combining a couple attributes, and I see both LLVM_ATTRIBUTE_UNUSED and [[gnu::unused]] in use. There's even [[fallthrough]] and [[clang::fallthrough]]. I'm not understanding what standard you see.

Reply:
Yeah, I'll follow up in discord about how we want to handle these.